File: app/src/main/java/adapters/NewsAdapter.java
package adapters;

import android.app.Activity;
import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.sqube.tipshub.NewsStoryActivity;
import com.sqube.tipshub.R;

import java.util.ArrayList;
import java.util.HashMap;

public class NewsAdapter extends RecyclerView.Adapter<NewsAdapter.ListNewsViewHolder>{
    private Activity activity;
    private ArrayList<HashMap<String, String>> data;

    public NewsAdapter(Activity a, ArrayList<HashMap<String, String>> d) {
        activity = a;
        data=d;
    }

    @NonNull
    @Override
    public ListNewsViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View convertView =  null;
        if(viewType==0)
            convertView = LayoutInflater.from(parent.getContext()).inflate(
                    R.layout.item_news, parent, false);
        else
            convertView = LayoutInflater.from(parent.getContext()).inflate(
                    R.layout.item_news_small, parent, false);

        return new ListNewsViewHolder(convertView);
    }

    @Override
    public int getItemViewType(int position) {
        return (position % 4);
    }

    @Override
    public void onBindViewHolder(@NonNull ListNewsViewHolder holder, int position) {
        HashMap<String, String> news = data.get(position);
        try{
            holder.description.setText(news.get("description"));
            holder.title.setText(news.get("title"));
            holder.time.setText(news.get("publishedAt"));

            if(news.get("urlToImage").length() > 5)
                Glide.with(activity)
                        .load(news.get("urlToImage"))
                        .into(holder.galleryImage);

            holder.crdContainer.setOnClickListener(v -> {
                Intent i = new Intent(activity.getApplicationContext(), NewsStoryActivity.class);
                i.putExtra("url", news.get("url"));
                activity.startActivity(i);
                activity.overridePendingTransition(R.anim.slide_right_in, R.anim.slide_left_out);
            });
        }catch(Exception e) {}

    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    class ListNewsViewHolder extends RecyclerView.ViewHolder {
        CardView crdContainer;
        ImageView galleryImage;
        TextView description, title, time;
        ListNewsViewHolder(View itemView) {
            super(itemView);
            crdContainer = itemView.findViewById(R.id.crdContainer);
            galleryImage = itemView.findViewById(R.id.galleryImage);
            description = itemView.findViewById(R.id.txtDescription);
            title = itemView.findViewById(R.id.title);
            time = itemView.findViewById(R.id.time);
        }
    }
}

File: app/src/main/java/adapters/NewsAdapter.kt
package adapters

import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import com.sqube.tipshub.R
import com.bumptech.glide.Glide
import android.net.Uri
import android.view.View
import android.widget.ImageView
import androidx.cardview.widget.CardView
import android.widget.TextView
import androidx.browser.customtabs.CustomTabsIntent
import com.sqube.tipshub.models.Article

class NewsAdapter(private val data: List<Article>) : RecyclerView.Adapter<NewsAdapter.ListNewsViewHolder>() {
    private val customTab = CustomTabsIntent.Builder()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListNewsViewHolder {
        val convertView: View = if (viewType == 0)
            LayoutInflater.from(parent.context).inflate(R.layout.item_news, parent, false)
        else
            LayoutInflater.from(parent.context).inflate(R.layout.item_news_small, parent, false)
        return ListNewsViewHolder(convertView)
    }

    override fun getItemViewType(position: Int): Int = position % 4

    override fun onBindViewHolder(holder: ListNewsViewHolder, position: Int) {
        val news = data[position]
        val customTabIntent = customTab.setToolbarColor(holder.itemView.context.resources.getColor(R.color.colorPrimary)).build()
        with(news){
            holder.description.text = description
            holder.title.text = title
            holder.time.text = publishedAt
            if (urlToImage != null) Glide.with(holder.galleryImage.context).load(urlToImage).into(holder.galleryImage)
            holder.crdContainer.setOnClickListener { customTabIntent.launchUrl(holder.itemView.context, Uri.parse(url)) }
        }
    }

    override fun getItemCount(): Int = data.size

    inner class ListNewsViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val crdContainer: CardView = itemView.findViewById(R.id.crdContainer)
        val galleryImage: ImageView = itemView.findViewById(R.id.galleryImage)
        val description: TextView = itemView.findViewById(R.id.txtDescription)
        val title: TextView = itemView.findViewById(R.id.title)
        val time: TextView = itemView.findViewById(R.id.time)
    }
}